File: C++/Day15.cpp
#include "AH.h"

namespace Day15
{

    typedef std::pair<std::string, int> P;
    
    std::tuple<std::string, bool, int> parseInput(const std::string & s)
    {
        auto upto = s.find_first_of("-=");
        bool flag = (s.at(upto) == '=');
        int value = -1;

        if (flag) {
            value = stoi(s.substr(upto + 1));
        }

        return { s.substr(0, upto), flag, value };
    }

    int hash(const std::string & s)
    {
        int hash = 0;

        for (auto c : s) {
            hash += int(c);
            hash *= 17;
            hash %= 256;
        }

        return hash;
    }

    int scoreBoxes(const std::vector<std::list<P>> boxes)
    {
        int score = 0;
        int b = 1;
        for (auto box : boxes) {
            int i = 1;
            for (auto l :box) {
                score += b * i * l.second;
                i++;
            }
            b++;
        }

        return score;
    }

	int Run(const std::string& filename)
	{
		const std::vector<std::string> ls = AH::ReadTextFile(filename);
        const std::vector<std::string> inputs = AH::Split(ls[0], ',');

		int part1 = 0;
        std::vector<std::list<P>> boxes(256);

        for (auto i : inputs) {
            const auto [ label, add, value ] = parseInput(i);
            const auto box_id = hash(label);
            part1 += hash(i);

            if (add) { // add to this box
                const P contents{label, value};
                // scan through the box to find a lens with the same label
                bool replaced = false;
                for (auto & l : boxes[box_id]) {
                    if (label == l.first) {
                        l.second = value;
                        replaced = true;
                    }
                }
                if (!replaced) {
                    boxes[box_id].push_back(contents);
                }
            } else { // remove from this box
                boxes[box_id].remove_if([label](P p){ return p.first == label; });
            }
        }

		AH::PrintSoln(15, part1, scoreBoxes(boxes));

		return 0;
	}

}

File: Go/Day15.go
package main

import (
	"strconv"
	"strings"

	AH "./adventhelper"
)

type P struct {
	s string
	n int
}

func parseInput(s string) (string, bool, int) {
	eq := strings.Index(s, "=")
	if eq > 0 {
		value, _ := strconv.Atoi(s[eq+1:])
		return s[:eq], true, value
	} else {
		return AH.TrimLastRune(s), false, 0
	}
}

func hash(s string) int {
	hash := 0

	for _, r := range s {
		hash += int(r)
		hash *= 17
		hash %= 256
	}

	return hash
}

func dropBox(boxes []P, label string) []P {
	n := -1
	for i, b := range boxes {
		if b.s == label {
			n = i
			break
		}
	}

	if n == -1 {
		return boxes
	} else {
		return append(boxes[:n], boxes[n+1:]...)
	}
}

func scoreBoxes(boxes [][]P) int {
	score := 0
	for b, box := range boxes {
		for i, l := range box {
			score += (b + 1) * (i + 1) * l.n
		}
	}

	return score
}

func main() {
	ls, _ := AH.ReadStrFile("../input/input15.txt")
	inputs := strings.Split(ls[0], ",")

	boxes := make([][]P, 256)
	part1 := 0

	for _, i := range inputs {
		label, add, value := parseInput(i)
		boxId := hash(label)
		part1 += hash(i)

		if add {
			replaced := false
			for i, l := range boxes[boxId] {
				if label == l.s {
					boxes[boxId][i].n = value
					replaced = true
					break
				}
			}
			if !replaced {
				boxes[boxId] = append(boxes[boxId], P{s: label, n: value})
			}
		} else {
			boxes[boxId] = dropBox(boxes[boxId], label)
		}
	}

	AH.PrintSoln(15, part1, scoreBoxes(boxes))

	return
}
